golem-worker-executor-base/src/services/rdbms/mod.rs

```diff
@@ -0,0 +1,741 @@
+
+    pub fn from(address: &str) -> Result<Self, String> {
+        let url = Url::parse(address).map_err(|e| e.to_string())?;
+        Ok(Self::new(url))
+    }
+
+    pub fn masked_address(&self) -> String {
+        let mut output: String = self.address.scheme().to_string();
+        output.push_str("://");
+
+        let username = self.address.username();
+        output.push_str(username);
+
+        let password = self.address.password();
+        if password.is_some() {
+            output.push_str(":*****");
+        }
+
+        if let Some(h) = self.address.host_str() {
+            if !username.is_empty() || password.is_some() {
+                output.push('@');
+            }
+
+            output.push_str(h);
+
+            if let Some(p) = self.address.port() {
+                output.push(':');
+                output.push_str(p.to_string().as_str());
+            }
+        }
+
+        output.push_str(self.address.path());
+
+        let query_pairs = self.address.query_pairs();
+
+        if query_pairs.count() > 0 {
+            output.push('?');
+        }
+        for (index, (key, value)) in query_pairs.enumerate() {
+            let key = &*key;
+            output.push_str(key);
+            output.push('=');
+
+            if key == "password" || key == "secret" {
+                output.push_str("*****");
+            } else {
+                output.push_str(&value);
+            }
+            if index < query_pairs.count() - 1 {
+                output.push('&');
+            }
+        }
+
+        output
+    }
+}
+
+impl Display for RdbmsPoolKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.masked_address())
+    }
+}
+
+impl IntoValue for RdbmsPoolKey {
+    fn into_value(self) -> Value {
+        Value::Record(vec![self.address.to_string().into_value()])
+    }
+
+    fn get_type() -> AnalysedType {
+        analysed_type::record(vec![analysed_type::field("address", analysed_type::str())])
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Encode, Decode)]
+pub struct DbRow<T: 'static> {
+    pub values: Vec<T>,
+}
+
+impl<T> RdbmsIntoValueAndType for DbRow<T>
+where
+    Vec<T>: RdbmsIntoValueAndType,
+{
+    fn into_value_and_type(self) -> ValueAndType {
+        let v = RdbmsIntoValueAndType::into_value_and_type(self.values);
+        let t = analysed_type::record(vec![analysed_type::field("values", v.typ)]);
+        ValueAndType::new(Value::Record(vec![v.value]), t)
+    }
+
+    fn get_base_type() -> AnalysedType {
+        analysed_type::record(vec![analysed_type::field(
+            "values",
+            <Vec<T>>::get_base_type(),
+        )])
+    }
+}
+
+impl<T> AnalysedTypeMerger for DbRow<T>
+where
+    T: AnalysedTypeMerger,
+    Vec<T>: RdbmsIntoValueAndType,
+{
+    fn merge_types(first: AnalysedType, second: AnalysedType) -> AnalysedType {
+        if let (AnalysedType::Record(f), AnalysedType::Record(s)) = (first.clone(), second) {
```

Comment:
what's the need of this merge_types?

Reply:
postgres have support for custom (user defined types),  where types are in general defined by data, this trait is used to create `AnalysedType` from basic and user defined types